src/core/public/workspace/workspaces_service.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { BehaviorSubject, combineLatest } from 'rxjs';
+import { isEqual } from 'lodash';
+
+import { CoreService, WorkspaceAttribute } from '../../types';
+
+type WorkspaceObject = WorkspaceAttribute & { readonly?: boolean };
+
+interface WorkspaceObservables {
+  /**
+   * Indicates the current activated workspace id, the value should be changed every time
+   * when switching to a different workspace
+   */
+  currentWorkspaceId$: BehaviorSubject<string>;
+
+  /**
+   * The workspace that is derived from `currentWorkspaceId` and `workspaceList`, if
+   * `currentWorkspaceId` cannot be found from `workspaceList`, it will return an error
+   *
+   * This value MUST NOT set manually from outside of WorkspacesService
```

Comment:
Yes, I think we can enforce it, that requires to refactor the current implementation a bit. Since this is the first PR, there will be other PRs will depends on this one, to avoid impacting others, what do you think if I create a task and come back to refactor this later?

Reply:
We can definitely do it in a follow-up; the concern here is to make sure we don't expose any public interfaces that we don't intend to, because those will all be subject to semver once they go out in a release.